canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+}  // namespace amazonaws
+}  // namespace com;
+
+// add frame pts, frame index, original frame size, CRC to beginning of buffer
+VOID addCanaryMetadataToFrameData(PFrame pFrame)
+{
+    PBYTE pCurPtr = pFrame->frameData;
+    putUnalignedInt64BigEndian((PINT64) pCurPtr, pFrame->presentationTs / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+    pCurPtr += SIZEOF(UINT64);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, pFrame->index);
+    pCurPtr += SIZEOF(UINT32);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, pFrame->size);
+    pCurPtr += SIZEOF(UINT32);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, COMPUTE_CRC32(pFrame->frameData, pFrame->size));
+}
+
+VOID createCanaryFrameData(PFrame pFrame)
+{
+    UINT32 i;
+
+    for (i = CANARY_METADATA_SIZE; i < pFrame->size; i++) {
+        pFrame->frameData[i] = RAND();
+    }
+    addCanaryMetadataToFrameData(pFrame);
+}
+
+void create_kinesis_video_frame(Frame *frame, const nanoseconds &pts, const nanoseconds &dts, FRAME_FLAGS flags,
+                                void *data, size_t len) {
+    frame->flags = flags;
+    frame->decodingTs = static_cast<UINT64>(dts.count()) / DEFAULT_TIME_UNIT_IN_NANOS;
+    frame->presentationTs = static_cast<UINT64>(pts.count()) / DEFAULT_TIME_UNIT_IN_NANOS;
+    // set duration to 0 due to potential high spew from rtsp streams
+    frame->duration = 0;
+    frame->size = static_cast<UINT32>(len);
+    frame->frameData = reinterpret_cast<PBYTE>(data);
+    frame->trackId = DEFAULT_TRACK_ID;
+}
+
+void updateFragmentEndTimes(UINT64 curKeyFrameTime, uint64_t &lastKeyFrameTime, map<uint64_t, uint64_t> *mapPtr)
+{
+        if (lastKeyFrameTime != 0)
+        {
+            (*mapPtr)[lastKeyFrameTime / HUNDREDS_OF_NANOS_IN_A_MILLISECOND] = curKeyFrameTime / HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+            cout << "Map Debug: current time is              " << duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count() << endl;
+            cout << "Map Debug: adding lastKeyFrameTime key: " << lastKeyFrameTime / HUNDREDS_OF_NANOS_IN_A_MILLISECOND << endl;
+            cout << "Map Debug: with curKeyFrameTime value:  " << curKeyFrameTime / HUNDREDS_OF_NANOS_IN_A_MILLISECOND<< endl;
+            auto iter = mapPtr->begin();
+            while (iter != mapPtr->end()) {
+                // clean up map: removing timestamps older than 5 min from now
+                if (iter->first < (duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count() - (300000)))
+                {
+                    iter = mapPtr->erase(iter);
+                    cout << "Map Debug: erasing a map key-value pair" << endl;
+                } else {
+                    break;
+                }
+            }
+        }
+        lastKeyFrameTime = curKeyFrameTime;
+}
+
+void pushKeyFrameMetrics(Frame frame, CustomData *cusData)
+{
+    updateFragmentEndTimes(frame.presentationTs, cusData->lastKeyFrameTime, cusData->timeOfNextKeyFrame);
+    
```

Comment:
Can we modularize this function further to get stream metrics and client metrics respectively? It will be cleaner.

Reply:
Separated client from stream, and I also modularized error metrics push into its own function too.